ledger/acctupdates.go

```diff
@@ -655,11 +552,87 @@ func (au *accountUpdates) getCatchpointStream(round basics.Round) (io.ReadCloser
+// not requiring any external locks, we can safely take a trackers lock on the ledger during reloadLedger, which ensures
+// that even during catchpoint catchup mode switch, we're still correctly protected by a mutex.
+type accountUpdatesLedgerEvaluator struct {
+	// au is the associated accountUpdates structure which invoking the trackerEvalVerified function, passing this structure as input.
+	// the accountUpdatesLedgerEvaluator would access the underlying accountUpdates function directly, bypassing the balances mutex lock.
+	au *accountUpdates
+	// prevHeader is the previous header to the current one. The usage of this is only in the context of initializeCaches where we iteratively
+	// building the StateDelta, which requires a peek on the "previous" header information.
+	prevHeader bookkeeping.BlockHeader
+}
+
+// GenesisHash returns the genesis hash
+func (aul *accountUpdatesLedgerEvaluator) GenesisHash() crypto.Digest {
+	return aul.au.ledger.GenesisHash()
+}
+
+// BlockHdr returns the header of the given round. When the evaluator is running, it's only referring to the previous header, which is what we
+// are providing here. Any attempt to access a different header would get denied.
+func (aul *accountUpdatesLedgerEvaluator) BlockHdr(r basics.Round) (bookkeeping.BlockHeader, error) {
+	if r == aul.prevHeader.Round {
+		return aul.prevHeader, nil
+	}
+	return bookkeeping.BlockHeader{}, ErrNoEntry{}
+}
+
+// Lookup returns the account balance for a given address at a given round
+func (aul *accountUpdatesLedgerEvaluator) Lookup(rnd basics.Round, addr basics.Address) (basics.AccountData, error) {
+	return aul.au.lookupImpl(rnd, addr, true)
+}
+
+// Totals returns the totals for a given round
+func (aul *accountUpdatesLedgerEvaluator) Totals(rnd basics.Round) (AccountTotals, error) {
+	return aul.au.totalsImpl(rnd)
+}
+
+// isDup return whether a transaction is a duplicate one. It's not needed by the accountUpdatesLedgerEvaluator and implemeted as a stub.
+func (aul *accountUpdatesLedgerEvaluator) isDup(config.ConsensusParams, basics.Round, basics.Round, basics.Round, transactions.Txid, txlease) (bool, error) {
+	// this is a non-issue since this call will never be made on non-validating evaluation
+	return false, fmt.Errorf("accountUpdatesLedgerEvaluator: tried to check for dup during accountUpdates initilization ")
+}
+
+// LookupWithoutRewards returns the account balance for a given address at a given round, without the reward
+func (aul *accountUpdatesLedgerEvaluator) LookupWithoutRewards(rnd basics.Round, addr basics.Address) (basics.AccountData, error) {
+	return aul.au.lookupImpl(rnd, addr, false)
+}
+
+// GetCreatorForRound returns the asset/app creator for a given asset/app index at a given round
+func (aul *accountUpdatesLedgerEvaluator) GetCreatorForRound(rnd basics.Round, cidx basics.CreatableIndex, ctype basics.CreatableType) (creator basics.Address, ok bool, err error) {
+	return aul.au.getCreatorForRoundImpl(rnd, cidx, ctype)
```

Comment:
no need to acquire the lock here?

Reply:
the `accountUpdatesLedgerEvaluator` function are called *only* in the following path:

loadFromDisk-> initializeFromDisk -> <evaluator> -> accountUpdatesLedgerEvaluator  :: GetCreatorForRound/Lookup/Totals

you'll see that in `loadFromDisk` we already took the lock, which is why we want to bypass taking the lock on all the `accountUpdatesLedgerEvaluator` functions.

Reply:
Now I understand better. TY.
Can we have it getCreatorForRound (not exported?)

Reply:
The function name need to match the one on the interface, which has a capital letter.
Note that *all* the accountUpdatesLedgerEvaluator belongs to the interface.

Reply:
Can't we make the interface lower letter? It looks like all are only accesed from within the package.

Reply:
Changing `ledgerForEvaluator` interface, also means that we will need to change the `Ledger` implementation, which is beyond what we're trying to achieve here since it will also force creating private/public interface for all these functions in the Ledger struct itself.